### ISSUE TITLE
Failed client authentication does not return an OAuth 2.0 error

### ISSUE THREAD

**Issue by michaeltecourt** [spring-projects/spring-security-oauth#603]
When providing bad client credentials (Authorization: Basic) to the /oauth/token endpoint, the error returned does not follow the OAuth 2.0 specification.
Example with Spring Boot :
```json
{"timestamp":1445441285803,"status":401,"error":"Unauthorized","message":"Bad credentials","path":"/oauth/token"}
```

I thought it should have the same format as when no credentials are provided : 
```json
{"error":"unauthorized","error_description":"Full authentication is required to access this resource"}
```

At first I thought it was an omission, but before going further with a fix, **I wanted to make sure this was not made on purpose** ?

[This incomplete patch](https://github.com/michaeltecourt/spring-security-oauth/commit/09c295c8a6ef14525e1308be2f59735f90e603fd) fixes the error format, but breaks integration tests : Response header `WWW-Authenticate` is now "Bearer..." instead of "Basic...".
A new `AuthenticationEntryPoint` implementation inspired by `OAuth2AuthenticationEntryPoint`, with the correct `WWW-Authenticate` response header, would do the math.

This issue can also be mitigated by using Spring Boot and a custom `ErrorController` that applies the correct error format.

**Comment by dsyer** [comment]
It's not really an OAuth2 authentication error, if the client fails to authenticate correctly, so I think the handler is the correct one (your patch adds the *wrong* handler for sure, as the integration test shows).

**Comment by michaeltecourt** [comment]
Yes, the header is definitely wrong as the patch is incomplete. I was halfway into implementing a new `AuthenticationEntryPoint` with the OAuth 2.0 error format (error, error_description) *and* the right `Basic ...` header, then I realized that maybe you did it on purpose. 

Technically client authentication errors are not described in the [spec](https://tools.ietf.org/html/rfc6749#section-4.1.2.1). It just felt weird to have two different models for the same endpoint, especially when these two errors are closely related from a client perspective : No credentials (OAuth 2.0 format) vs Bad credentials (other format).

**Comment by dsyer** [comment]
I guess so (weirdness). Are you going to update the PR with a "Basic ..." header and the "correct " body?

**Comment by michaeltecourt** [comment]
Well if you find it helpful I can finish it... depends on you :sweat_smile:
If you're not interested our team's project relies on the fallback option mentioned above (Spring Boot `ErrorController`), so it's definitely not critical, even to nitpickers.

**Comment by kumarabhishek0809** [comment]
I downgraded my spring boot version and it worked.
springBootVersion = '1.4.3.RELEASE'
springCloudVersion = 'Camden.SR3'
